pkg/controller/controller_test.go

```diff
@@ -3156,3 +3129,257 @@ func TestProvisionFromPVC(t *testing.T) {
+			expectErr:   true,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Set up test
+			tmpdir := tempDir(t)
+			defer os.RemoveAll(tmpdir)
+			mockController, driver, _, controllerServer, csiConn, err := createMockServer(t, tmpdir)
+			if err != nil {
+				t.Fatal(err)
+			}
+			mockTranslator := NewMockProvisionerCSITranslator(mockController)
+			defer mockController.Finish()
+			defer driver.Stop()
+			clientSet := fakeclientset.NewSimpleClientset()
+			pluginCaps, controllerCaps := provisionCapabilities()
+			csiProvisioner := NewCSIProvisioner(clientSet, 5*time.Second, "test-provisioner",
+				"test", 5, csiConn.conn, nil, driverName, pluginCaps, controllerCaps,
+				inTreePluginName, false, mockTranslator)
+
+			// Set up return values (AnyTimes to avoid overfitting on implementation)
+			if tc.expectTranslation {
+				// Have fake translation provide same as input SC but with
+				// Parameter indicating it has been translated
+				mockTranslator.EXPECT().TranslateInTreeStorageClassToCSI(gomock.Any(), gomock.Any()).DoAndReturn(
+					func(inTreePluginName string, sc *storagev1.StorageClass) (*storagev1.StorageClass, error) {
+						newSC := sc.DeepCopy()
+						newSC.Parameters[translatedAnnKey] = "foo"
+						return newSC, nil
+					},
+				).AnyTimes()
+				// Have fake translation provide same as input PV but with
+				// Annotation indicating it has been translated
+				mockTranslator.EXPECT().TranslateCSIPVToInTree(gomock.Any()).DoAndReturn(
+					func(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+						newPV := pv.DeepCopy()
+						if newPV.Annotations == nil {
+							newPV.Annotations = map[string]string{}
+						}
+						newPV.Annotations[translatedAnnKey] = "foo"
+						return newPV, nil
+					},
+				).AnyTimes()
+			}
+
+			// Set an expectation that the Create should be called
+			if tc.expectCreate {
+				expectParams := map[string]string{"fstype": "ext3"} // Default
+				if tc.expectTranslation {
+					// If translation is expected we check that the CreateVolume
+					// is called on the expected volume with a translated param
+					expectParams[translatedAnnKey] = "foo"
+				}
+				controllerServer.EXPECT().CreateVolume(gomock.Any(),
+					&csi.CreateVolumeRequest{
+						Name:               "test-testi",
+						Parameters:         expectParams,
+						VolumeCapabilities: nil,
+						CapacityRange: &csi.CapacityRange{
+							RequiredBytes: int64(requestBytes),
+						},
+					}).Return(
+					&csi.CreateVolumeResponse{
+						Volume: &csi.Volume{
+							CapacityBytes: requestBytes,
+							VolumeId:      "test-volume-id",
+						},
+					}, nil).Times(1)
+			}
+
+			// Make a Provision call
+			volOpts := controller.ProvisionOptions{
+				StorageClass: &storagev1.StorageClass{
+					Provisioner:   tc.provisioner,
+					Parameters:    map[string]string{"fstype": "ext3"},
+					ReclaimPolicy: &deletePolicy,
```

Comment:
Is this one of those APIs where you can't take the address of a literal, so you need to either declare the value beforehand or wrap it in a function that returns a pointer? :cry:

Reply:
yeah